### ISSUE TITLE
Possible bug with `correlate()` method?

### ISSUE THREAD

**Issue by raaaaaymond** [dtschump/CImg#255]
I've already posted this on Stack Overflow [here](https://stackoverflow.com/questions/60949377/cimg-possible-bug-in-image-correlation). I'll repeat it below.

Windows 10, Visual Studio 2019, CImg commit `ac3195eb5dc75c9b8efc95e9fde2f1a0ea3b77cd`.

There seems to be a bug with the `get_correlate()` method of the `CImg` class. For convenience, I paste the signature below, according to the [official documentation](http://cimg.eu/reference/structcimg__library_1_1CImg.html#a6ae0634808824131083214f2a5a518ca).

```cpp
CImg<T>& correlate	(
  const CImg< t > & 	kernel,
  const unsigned int 	boundary_conditions = 1,
  const bool 	is_normalized = false,
  const unsigned int 	channel_mode = 1,
  const unsigned int 	xcenter = ~0U,
  const unsigned int 	ycenter = ~0U,
  const unsigned int 	zcenter = ~0U,
  const unsigned int 	xstart = 0,
  const unsigned int 	ystart = 0,
  const unsigned 	zstart = 0,
  const unsigned int 	xend = ~0U,
  const unsigned int 	yend = ~0U,
  const unsigned int 	zend = ~0U,
  const float 	xstride = 1,
  const float 	ystride = 1,
  const float 	zstride = 1,
  const float 	xdilation = 1,
  const float 	ydilation = 1,
  const float 	zdilation = 1
)
```

Here, `~0U` means "centred". (Note that this is the `correlate()` method; `get_correlate()` is the same thing but gives you a new instance instead of modifying in-place.)

To reproduce the bug, use the simple example below.

```cpp
#include <iostream>
#include <CImg.h>

using cimg_library::CImg;

int main() {
  // Create a 20x20 image and draw a white square in the middle.
  CImg<unsigned int> image {20, 20, 1, 1};
  image.fill(50);
  image.draw_rectangle(5, 5, 0, 0, 14, 14, 0, 0, 200);

  // Create a simple kernel of just ones.
  CImg<double> kernel {3, 3, 1 ,1};
  kernel.fill(1.0);

  // Bug? Below, if I set the centre to be the actual centre (1, 1) of the kernel, the resulting correlated image seems
  // to be all zeros.
  CImg correlated = image.get_correlate(kernel, 1, false, 1, 1, 2, 0);

  // Just for debugging.
  std::cout << "Max: " << correlated.max() << ", min: " << correlated.min() << ".\n";

  // Save and display the correlated image.
  correlated.save("correlated.png");
  correlated.display();
  return 0;
}
```

The constructed `image` looks like [this](https://i.stack.imgur.com/GSYj2.png)

The code above is not yet buggy as-is. The result, `correlated`, is correctly computed, and it looks like [this](https://i.stack.imgur.com/Yrfpv.png).

However, if instead of the line `CImg correlated = image.get_correlate(kernel, 1, false, 1, 1, 2, 0)`, we do `CImg correlated = image.get_correlate(kernel, 1, false, 1, 1, 1, 0)`, which means to set the centre correlation to the actual centre, (1, 1), of the kernel, then we get a garbage result where all the pixels are zero or near-zero (on my machine anyway). This seems like a bug to me. Can anyone else reproduce it? This is a weird but, because I would have thought the most common usage of correlation would be to set the centre of correlation to the actual centre of the kernel.

**Comment by raaaaaymond** [comment]
I just did some very brief debugging. 

There is a special case in function `_correlate()` for *centred* 3x3 kernels, and the result is being assigned on line 36477. However, what is being assigned there is not `res`, but `_res`, which is supposed to be a shared memory reference of `res`, but upon checking the address of `_res.data`, it is different from `res.data`, and so maybe the result is being assigned to the wrong address, and so `res` still has uninitialised garbage in it at the time it is returned?

For non-centred kernels, `res` is being assigned directly on line 36635, instead of through a reference `_res`, and we see that the correct result is returned.

So in summary, the problem could be that the method `get_shared_channel` isn't getting the actual shared channel? In any case, the computed numbers are being assigned to the wrong address.

Incidentally, why are you using a shared memory reference `_res` for the special case of centred 3x3 kernels anyway? Why not just assign to `res` directly?

**Comment by raaaaaymond** [comment]
Another update, I've tried a kernel of 5x5 and the bug still exists, but it works fine for a kernel of size 7x7. Looking at the code, this is not surprising because both 3x3 and 5x5 kernels are treated specially.

**Comment by dtschump** [comment]
I can confirm this.
Currently trying to understand what is going on. Thanks for the report!

**Comment by dtschump** [comment]
This seems to be related to wrong management of CImg datatypes.
If you define `image` to be a `CImg<double>`, then it works as expected.

**Comment by raaaaaymond** [comment]
Yep can confirm, changing `image` to `CImg<double>` works.

**Comment by dtschump** [comment]
I think I've found the bug.
This commit should fix it:
https://github.com/dtschump/CImg/commit/6f4a764b9b0a0654d13d6e3750ec1935f76c28f2

I'm releasing a new CImg package with the fix. Should be available in a few minutes on the website.

**Comment by raaaaaymond** [comment]
Oh wow that's amazing thank you very much! That was lightning fast.
